src/locator.rs

```diff
@@ -0,0 +1,643 @@
+                None
+            } else {
+                Some(s.to_string())
+            }
+        });
+
+        match parse_org_project(&project) {
+            Ok((org_id @ Some(_), project)) => Ok(Locator {
+                fetcher,
+                org_id,
+                project: String::from(project),
+                revision,
+            }),
+            Ok((org_id @ None, _)) => Ok(Locator {
+                fetcher,
+                org_id,
+                project,
+                revision,
+            }),
+            Err(error) => Err(Error::Parse(ParseError::Project {
+                input: locator.to_owned(),
+                project,
+                error,
+            })),
+        }
+    }
+
+    /// Promote a `Locator` to a [`StrictLocator`] by providing the default value to use
+    /// for the `revision` component, if one is not specified in the locator already.
+    ///
+    /// The `ToString` implementation is lazily evaluated if the locator doesn't already contain a revision.
+    pub fn promote_strict(self, revision: impl ToString) -> StrictLocator {
+        let locator = StrictLocator::builder()
+            .fetcher(self.fetcher)
+            .project(self.project)
+            .revision(self.revision.unwrap_or_else(|| revision.to_string()));
+
+        match self.org_id {
+            None => locator.build(),
+            Some(org_id) => locator.org_id(org_id).build(),
+        }
+    }
+
+    /// Promote a `Locator` to a [`StrictLocator`] by providing the default value to use
+    /// for the `revision` component, if one is not specified in the locator already.
+    ///
+    /// The revision is lazily evaluated if the locator doesn't already contain a revision.
+    pub fn promote_strict_with<F: Fn() -> String>(self, revision: F) -> StrictLocator {
+        let locator = StrictLocator::builder()
+            .fetcher(self.fetcher)
+            .project(self.project)
+            .revision(self.revision.unwrap_or_else(revision));
+
+        match self.org_id {
+            None => locator.build(),
+            Some(org_id) => locator.org_id(org_id).build(),
+        }
+    }
+
+    /// Converts the locator into a [`PackageLocator`] by discarding the `revision` component.
+    /// Equivalent to the `From` implementation, but offered as a method for convenience.
+    pub fn into_package(self) -> PackageLocator {
+        self.into()
+    }
+
+    /// Explodes the locator into its (owned) parts.
+    /// Used for conversions without cloning.
+    pub(crate) fn explode(self) -> (Fetcher, Option<usize>, String, Option<String>) {
+        (self.fetcher, self.org_id, self.project, self.revision)
+    }
+}
+
+impl Ord for Locator {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        match self.fetcher.cmp(&other.fetcher) {
+            Ordering::Equal => {}
+            ord => return ord,
+        }
+        match alphanumeric_sort::compare_str(&self.project, &other.project) {
```

Comment:
We're not ordering by org ID here, but I think we should be. If you change the order of the locators in the test on line 541 to this, then the test fails:

```
        let locators = vec![
            "git+github.com/foo/bar",
            "git+github.com/foo/bar$1234",
            "custom+baz$1234",
            "custom+2/bam$1234",
            "custom+1/bam$1234",
            "custom+2/bam",
        ]
```